### ISSUE TITLE
Feature request: Support a values file directory, such as values.d

### ISSUE THREAD

**Issue by tculp** [helm/helm#10416]
Values.yaml files often get large and can get unwieldy. It would be convenient to be able to easily split up values files.

While multiple values.yaml files can be specified on the command line, it is somewhat cumbersome, especially with more than one or two files. Also, in subcharts, additional values files are infeasible.

A common paradigm in the linux world is to have a combination of a config file and a config.d folder, such as for system services. Such a paradigm could be used by helm to allow a values.d or values.yaml.d folder, which would automatically get loaded and merged with the values.yaml file.

Even if it was a non-default parameter, such as --values-dir, this would still be quite useful.

**Comment by R-omk** [comment]
@tculp, as a workaround, you can create multiple subcharts just to split  values, although there are some issues with subcharts right now.

**Comment by tculp** [comment]
@hickeyma Thanks for removing the stale label, but the issue appears to have not been re-opened

**Comment by hickeyma** [comment]
Sorry forgot that. Re-opened now.

**Comment by tculp** [comment]
Un-stale

**Comment by tculp** [comment]
Un-stale

**Comment by ShivamTyagi12345** [comment]
This is interesting. I will take this 

/ assign

**Comment by Bhargav-InfraCloud** [comment]
@tculp @joejulian @hickeyma Can you please review #11626 if it works as expected? Thanks!

**Comment by tculp** [comment]
@Bhargav-InfraCloud The implementation is a little strange. The paradigm of automatically loading files in the directory based on the supplied -f flag is somewhat unwieldy, and requires a values file with the exact name as exists in the values.d folder. E.g. for values.d/values2.yaml/extra.yaml to be loaded, I need to have a values2.yaml to pass to the -f flag. Also, it looks like the extra.yaml is hardcoded, so you basically can only split one file into 2.

I think a better implementation would be to be able to by default load all recursive files under the values.d folder, and to be able to specify additional folder with a new flag like '-d, --values-directory'.

For example:
```
values.yaml # loaded automatically
values.d/
--global/
    --values.yaml # loaded automatically
--ingress.yaml
--deployment.yaml

values-extra.yaml # loaded with the -f flag
values-others.d/
--ingress.yaml # loaded with `-d values-others.d`
```

**Comment by Bhargav-InfraCloud** [comment]
Hey @tculp. Guess my `extra.yaml` caused some confusion (couldn't come up with a better name though). It is not handcoded for usage but for a sample to user. The idea is `helm create` will create the sample `./values.d/values.yaml/extra.yaml` along with other files like `values.yaml`, `charts/`, etc. But there is **no dependency on the name** as such.
For eg., all the following cases are **valid**:
#### 1. Without splitting values:
```
 mychart/
 ├── values.d/
 └── values.yaml
```
`Note`: The directory `values.d/` is totally optional.
#### 2. Original sample provided to user when running `helm create`
```
 mychart/
 ├── values.d/
 │   └── values.yaml/
 │       └── extra.yaml
 └── values.yaml
```

#### 3. With other value files
```
 mychart2/
 ├── values.d/
 │   └── values.yaml/
 │       ├── extra.yaml
 │       ├── values0.yaml         # Any names really
 │      ...
 │       └── valuesN.yaml
 └── values.yaml
```

#### 4. With other value files and w/o `extra.yaml`
```
 mychart2/
 ├── values.d/
 │   └── values.yaml/
 │       └── xyz.yaml            # Any name really
 └── values.yaml
```

---
About loading all the YAMLs from `values.d/` directly, my thoughts don't align right with it. I think users can also choose to classify the values files based on environment, etc., say `dev-values.yaml`, `prod-values.yaml`, etc. In cases similar to this, having multiple levels like `values.d/<input-values-filename>` would help classify (or) maintain the values better. Loading all files from `values.d/` can only help to certain extent.

A second approach here is to have multiple `values.d/`.
```
 mychart2/
 ├── dev-values.d/          # Match value's file's basename
 │   └── abc.yaml
 ├── prod-values.d/         # Match value's file's basename
 │   └── xyz.yaml
 ├── dev-values.yaml
 └── prod-values.yaml
```
Either approach is fine. I had to pick one. Your thoughts?

**Comment by tculp** [comment]
Hey @Bhargav-InfraCloud, I see that I was mistaken about the extra.yaml. I think that the second option (having multiple directories) makes more sense, and I think the paradigm of automatically loading a folder makes more sense then. Still, it might be better to just require directories to be called out with a flag rather than automatically picking them based on the filenames provided. I could see a case where people want just a folder and have to make an empty file to pass the -f flag to. Thoughts?

**Comment by Bhargav-InfraCloud** [comment]
@tculp  Guess I missed your actual request was to feed values as directory directly. My apologies. My initial thoughts were to only extend the existing values file, say `values.yaml`, to `values.d/values.yaml/*` or `values.d/*`. But now I think I'm fine with any of the approaches as long as the user gets to use both values file and directory.

@ everyone on this chat,
Since there are more than one approach discussed here, if you are following, please give your thoughts on them.
<table>
    <tr>
       <td>Approach no.</td>
       <td>Command (Partial)</td>
       <td>Idea</td>
       <td>Pros</td>
       <td>Cons</td>
    </tr>
    <tr>
       <td>1</td>
       <td>helm install --values-directory values.d</td>
       <td>Read all values files in `values.d/`</td>
       <td>1. No change in existing flag `-f` or `--values` 's behaviour.</td>
       <td>1. Need to introduce new command line flag `--values-directory`.</td>
    </tr>
    <tr>
       <td>2</td>
       <td rowspan="2">helm install --values values.yaml</td>
       <td>Read `values.yaml` and all files from `values.d/values.yaml/*`, if `values.d/values.yaml/` exists. Else, only `values.yaml`.</td>
       <td rowspan="2">1. Only extending the existing values to a directory.</p>2. No new flags are required.</td>
       <td rowspan="2">1. Adds extra logic to existing flag `-f` or `--values` to check if matching directory exists.</p>2. If only values directory have to be fed as input, the matching values file still needs to be created for the sake of input.</td>
    </tr>
    <tr>
       <td>3</td>
       <td>Similar to approach 2, but reads `values.yaml` and all files from directory which matching basename of input file, i.e., `values.d/*`, if `values.d/` exists. Else, only `values.yaml`.</td>
    </tr>
    <tr>
       <td>4</td>
       <td>helm install --values values.yaml</p>helm install --values values.d</td>
       <td>Combined approach 1 &amp; 3, i.e., read only values file if `values.yaml` is provided as input. And read all files in directory if `values.d` is provided as input.</td>
       <td>1. No new flags are required.</p>2. No need of adding values file if only directory is to be fed.</td>
       <td>1. Adds extra logic to existing flag `-f` or `--values` to check if matching directory exists.</td>
    </tr>
 </table>

**Comment by joejulian** [comment]
My personal opinions:

`#1` provides the least surprise and matches a method that is common throughout the industry.
`#2` is confusing and likely to cause mistakes.
`#3` is ok, still a bit likely to cause confusion and mistakes, imo.
`#4` changes the behavior of an existing flag which is technically considered a breaking change. It might be ok, but I still would prefer the explicit expectations that `--values` expects a file and `--values-directory` expects a directory.

**Comment by tculp** [comment]
I definitely prefer either `#1` or `#4`. `#1` is probably better for the reasons @joejulian mentioned, and is pretty simple to use if `--values-directory` also has a `-d` shorthand.

**Comment by joejulian** [comment]
> Also as @sirianni mentioned in the pull request, the plain values.d directory could be considered loaded by default, just like values.yaml

That'll probably be considered a breaking change by the core developers to change the default behavior. It's entirely possible (though probably not a very good idea) to have shipped example values under values.d/, for example, and changing that behavior would break any chart that did so.

**Comment by tculp** [comment]
@Bhargav-InfraCloud  I tested it out and it works well. A couple points, should the directory be loaded recursively? I noticed it currently only loads the first level of files. This is more a question for the larger group I suppose.

I also found that the directory values always override values files, regardless of order. For example, a `-f values.yaml -d values.d -f overridingvalues.yaml` will fail to use the overriding values, whereas `-f values.yaml -f overridingvalues.yaml` does override using the values in the later specified files.

**Comment by Bhargav-InfraCloud** [comment]
> I noticed it currently only loads the first level of files. This is more a question for the larger group I suppose.

Exactly my thoughts. I did add a comment for this in code thinking we can always come back to this. But let me add this and submit.

> I also found that the directory values always override values files, regardless of order.

Yep. That's how helm reads inputs currently. The precedence order is:

1. `--set-file`
2. `--set-string`
3. `--set`
4. `--set-json`
5. `-d/--values-directory`
6. `-f/--values`

The overriding happens if both inputs are of the same precedence or a higher precedence input type for the latter input. In your example: (`-f values.yaml -d values.d -f overridingvalues.yaml`), `values.d` can override `values.yaml` but `values.d` cannot be overridden by `overridingvalues.yaml`. `overridingvalues.yaml` can override `values.yaml` since both are of same priority.

In other words, inputs using `-f/--values` can be overridden by any input type including itself, but the opposite is not possible.

The only possible change here, if you think this is how it should be, we can make `-d/--values-directory` least in priority (i.e, swap 5 & 6).

**Comment by Bhargav-InfraCloud** [comment]
@tculp /all Made the 2 changes as discussed. Please validate.

**Comment by tculp** [comment]
@Bhargav-InfraCloud Tested it out, looks good to me!

**Comment by Bhargav-InfraCloud** [comment]
@tculp Great! Thanks for confirming 👍🏼 
@joejulian Can you please review or add `awaiting review` tag on this card or #11626 for other maintainers to review?

**Comment by SytemicAnomaly** [comment]
Hi !
I think it would be even better if the support for directories was included with the --values flag instead of a new --values-directory flag.
It wouldn't be a breaking change as it would not change anything for currently working configurations with filepaths and would only extend it to work if a directory is provided (which would have resulted in an error with the current release).

It would also be closer with how kubectl apply works with files and directories with -f and be more easily integrated with ArgoCD for example (it would work out-of-the box).

**Comment by Bhargav-InfraCloud** [comment]
Hey @Syst3m1cAn0maly, I just added a comment for it [here][1]. Please check.

[1]: https://github.com/helm/helm/pull/11626#issuecomment-1546856977

**Comment by ryan-dyer-sp** [comment]
@Bhargav-InfraCloud Any update on this functionality or what may be preventing the PR from moving out of Draft?

**Comment by Bhargav-InfraCloud** [comment]
@ryan-dyer-sp Apologies for the late reply. The PR was size XL so I wanted to make it smaller, if possible. Just did that. Please review.

**Comment by danielloader** [comment]
I hate bumping issue threads with a comment like "+1" so I'll put some effort in.

This would be really handy for unpleasantly large sub-chart deployments - I've inherited an installation at my company with a chart with 12 sub-charts, each with a lengthy values block. 

Deploying it in ArgoCD has necessitated just having a very large `values.{tenant}.yaml` file which I'd very much prefer to have `./{tenant}/**.yaml` merged into one values map and applied.

While I can abuse the `ignoreMissingValueFiles` and have a very large bank of values files explicitly configured and just skip ones that don't exist, I'd rather just merge an entire directory as is without trying to remember the values file semantic conventions in filenames in lockstep with any additional sub-charts added.

As such I am still cheerleading for this change!

**Comment by jan-mrm** [comment]
hmm, is there any motivation to add this feature - to get the offered PR https://github.com/helm/helm/pull/11626 merged? If so, what is needed to get there?
I'm personally a fan of this feature and it would be handy.
